connectors/testconn/connector.go

```diff
@@ -0,0 +1,383 @@
+		}
+		bsonBytes, err := bson.Marshal(res)
+		if err != nil {
+			return err
+		}
+		batch = append(batch, bsonBytes)
+	}
+	err = f.Close()
+	if err != nil {
+		return err
+	}
+
+	f2, err := os.Open(filepath.Join(c.path, "updates.json"))
+	if err != nil {
+		return err
+	}
+	defer f2.Close()
+
+	scanner2 := bufio.NewScanner(f2)
+	var allSplitted [][]string
+	for scanner2.Scan() {
+		line := strings.TrimSpace(scanner2.Text())
+		if line == "" {
+			continue
+		}
+		splitted := strings.SplitN(line, "\t", 3)
+		allSplitted = append(allSplitted, splitted)
+	}
+	err = f2.Close()
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	for _, task := range readPlan.Tasks {
+		id := task.Id
+		db := task.Def.Db
+		col := task.Def.Col
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			channel <- iface.DataMessage{
+				DataBatch:    batch,
+				MutationType: iface.MutationType_InsertBatch,
+				Loc: iface.Location{
+					Database:   db,
+					Collection: col,
+				},
+			}
+		Loop:
+			for {
+				for _, splitted := range allSplitted {
+					var idRes interface{}
+					// Weird, but wrapping the id this lets us parse the type properly
+					err := bson.UnmarshalExtJSON([]byte("{\"id\": "+splitted[1]+"}"), true, &idRes)
+					if err != nil {
+						slog.Error(err.Error())
+						break Loop
+					}
+					idType, idVal, err := bson.MarshalValue(idRes.(bson.D)[0].Value)
+
+					if err != nil {
+						slog.Error(err.Error())
+						break Loop
+					}
+
+					var data *[]byte
+					var mutationType uint
+
+					if splitted[0] == "delete" {
+						mutationType = iface.MutationType_Delete
+					} else if splitted[0] == "insert" {
+						mutationType = iface.MutationType_Insert
+						var res interface{}
+						err = bson.UnmarshalExtJSON([]byte(splitted[2]), true, &res)
+						if err != nil {
+							slog.Error(err.Error())
+							break Loop
+						}
+						doc, err := bson.Marshal(res)
+						if err != nil {
+							slog.Error(err.Error())
+							break Loop
+						}
+						data = &doc
+					} else if splitted[0] == "update" {
+						mutationType = iface.MutationType_Update
+						var res interface{}
+						err = bson.UnmarshalExtJSON([]byte(splitted[2]), true, &res)
+						if err != nil {
+							slog.Error(err.Error())
+							break Loop
+						}
+						doc, err := bson.Marshal(res)
+						if err != nil {
+							slog.Error(err.Error())
+							break Loop
+						}
+						data = &doc
+					}
+					select {
+					case <-c.flowCtx.Done():
+						return
+					case channel <- iface.DataMessage{
+						Data:         data,
+						MutationType: mutationType,
+						Loc: iface.Location{
+							Database:   db,
+							Collection: col,
+						},
+						Id:     &idVal,
+						IdType: byte(idType),
+					}:
+					}
+				}
+				if !c.loop {
+					break
+				}
+			}
+
+			err = c.coord.NotifyTaskDone(flowId, c.id, id, nil)
+			if err != nil {
+				slog.Error(err.Error())
+			}
+		}()
+	}
+
+	go func() {
+		wg.Wait()
+		c.coord.NotifyDone(flowId, c.id)
```

Comment:
Between Wait() and NotifyDone(), we should close the data channel to signal downstream that nothing else is coming their way